source/common/common/token_bucket_impl.cc

```diff
@@ -56,4 +67,52 @@ void TokenBucketImpl::maybeReset(uint64_t num_tokens) {
   last_fill_ = time_source_.monotonicTime();
 }
 
+AtomicTokenBucketImpl::AtomicTokenBucketImpl(uint64_t max_tokens, TimeSource& time_source,
+                                             double fill_rate, bool init_fill)
+    : max_tokens_(max_tokens), fill_rate_(std::max(std::abs(fill_rate), kMinFillRate)),
+      time_source_(time_source) {
+  auto time_in_nanoseconds = time_source.monotonicTime();
+  if (init_fill) {
+    time_in_nanoseconds -= toNanoseconds(max_tokens_ / fill_rate_);
+  }
+  time_in_nanoseconds_.store(time_in_nanoseconds);
+}
+
+uint64_t AtomicTokenBucketImpl::consume(uint64_t tokens, bool allow_partial, double factor) {
+  ASSERT(factor >= 0.0);
+  ASSERT(factor <= 1.0);
+
+  const MonotonicTime time_now = time_source_.monotonicTime();
+  MonotonicTime time_expected = time_in_nanoseconds_.load();
+  MonotonicTime time_new{};
+
+  const double used_fill_rate = fill_rate_ * factor;
+
+  do {
+    if (time_now <= time_expected) {
+      return 0;
+    }
+
+    const double total_tokens =
+        std::min(max_tokens_, toSeconds(time_now - time_expected) * used_fill_rate);
+
+    const uint64_t available_tokens = static_cast<uint64_t>(std::floor(total_tokens));
+    if (available_tokens < tokens) {
+      if (allow_partial) {
+        tokens = available_tokens;
+      } else {
+        return 0;
+      }
+    }
+
+    // Move the time_in_nanoseconds_ forward by the number of tokens consumed.
+    const double total_tokens_new = total_tokens - static_cast<double>(tokens);
+    time_new = time_now - toNanoseconds(total_tokens_new / used_fill_rate);
+
+    ASSERT(time_new >= time_expected);
+  } while (!time_in_nanoseconds_.compare_exchange_weak(time_expected, time_new));
```

Comment:
should the `load` be inside the opportunistic loop? looks odd to have load outside the hot loop.

Reply:
ah, `compare_exchange_weak` does reload. I noticed that local rate limit uses relaxed memory order, by the way.

Reply:
yeah.